toqito/matrices/tests/test_operator_sinkhorn.py

```diff
@@ -0,0 +1,179 @@
+        expected_msg = (
+            "If dim is of size 1, rho must be square and dim[0] must evenly divide length(rho); "
+            "please provide the dim array containing the dimensions of the subsystems."
+        )
+        assert str(e) == expected_msg
+
+
+def test_operator_sinkhorn_invalid_dim_array():
+    """Test operator Sinkhorn when product of the dim array does not match the density matrix dims."""
+    dX = 8
+    dim1 = [4, 3, 2]  # 4*3*2 != 8
+
+    rho = random_density_matrix(8)  # 8-dimensional density matrix
+
+    # The dimension `4` does not divide evenly into `8`, so we expect an error
+    try:
+        operator_sinkhorn(rho, dim=dim1)
+    except ValueError as e:
+        expected_msg = f"Product of dimensions {dim1} does not match rho dimension {dX}."
+        assert str(e) == expected_msg
+
+
+def test_operator_sinkhorn_valid_single_dim():
+    """Test operator Sinkhorn when a single valid number is passed as `dim`."""
+    rho = random_density_matrix(9)  # 9-dimensional density matrix
+
+    # The dimension `3` divides evenly into `9`, so no error should occur
+    sigma, F = operator_sinkhorn(rho, dim=[3])
+
+    # Check that sigma is a valid density matrix with trace equal to 1
+    np.testing.assert_almost_equal(np.trace(sigma), 1)
+
+
+def test_operator_sinkhorn_max_mixed():
+    """Test operator Sinkhorn on a maximally mixed bipartite state. Should be invariant."""
+    rho = np.eye(9)  # 9-dimensional density matrix
+
+    # The dimension `3` divides evenly into `9`, so no error should occur
+    sigma, F = operator_sinkhorn(rho, dim=[3])
+
+    # Check that rho is invariant after sinkhorn operation
+    np.testing.assert_almost_equal(sigma, rho)
+
+
+def test_operator_sinkhorn_max_entangled():
+    """Test operator Sinkhorn on a maximally entangled bipartite state. Should be invariant."""
+    # function should return the inintial state since it already satisfies the trace property
+
+    u0 = bell(0)
+    rho = u0 @ u0.conj().T
+
+    sigma, F = operator_sinkhorn(rho, dim=[2])
+
+    # Check that rho is invariant after sinkhorn operation
+    np.testing.assert_almost_equal(sigma, rho)
+
+
+def test_operator_sinkhorn_non_square_rho():
+    """Test operator sinkhorn on non-square input matrix."""
+    # function should raise a ValueError
+
+    rho = np.random.rand(4, 5)
+    try:
+        operator_sinkhorn(rho)
+    except ValueError as e:
+        expected_msg = "Input 'rho' must be a square matrix."
+        assert str(e) == expected_msg
+
+
+def test_operator_sinkhorn_max_iterations():
+    """Test operator sinkhorn on insufficient iteration limit."""
+    # function should raise a RuntimeError
+
+    rho_random = random_density_matrix(4, seed=42)
+    try:
+        operator_sinkhorn(rho=rho_random, dim=[2, 2], max_iterations=20)
+    except RuntimeError as e:
+        expected_msg = "operator_sinkhorn did not converge within 20 iterations."
+        assert str(e) == expected_msg
```

Comment:
most of these raises error tests can be condensed into one. If you would like an example, search through the existing unit tests. 

https://docs.pytest.org/en/7.1.x/how-to/parametrize.html